model-optimizer/extensions/ops/einsum.py

```diff
@@ -0,0 +1,248 @@
+                is_inputs_permuted.append(True)
+                labels.insert(1, labels[-1])
+                del labels[-1]
+            else:
+                is_inputs_permuted.append(False)
+            permuted_input_subscript = ''.join(labels)
+            permuted_input_subscripts.append(permuted_input_subscript)
+
+        # perform the same procedure for the output subscript as for the inputs subscripts
+        labels = Einsum.extract_subscript_labels(node_name, output_subscript)
+        num_broadcated_dims = output_rank - len(labels) + 1
+        if output_rank > 3 and labels[-1] != "...":
+            is_output_permuted = True
+            labels.insert(1, labels[-1])
+            del labels[-1]
+        elif output_rank > 3 and labels[-1] == "..." and num_broadcated_dims == 1:
+            is_output_permuted = True
+            labels.insert(1, labels[-1])
+            del labels[-1]
+        else:
+            is_output_permuted = False
+        permuted_output_subscript = ''.join(labels)
+
+        # concatenate the left and right hands of the resulted equation
+        left_hand = ','.join(permuted_input_subscripts)
+        right_hand = permuted_output_subscript
+        permuted_equation = left_hand + "->" + right_hand
+        return permuted_equation, is_inputs_permuted, is_output_permuted
+
+    @staticmethod
+    def infer(node: Node):
+        node_name = node.soft_get('name', node.id)
+        connected_in_ports = [port for port in node.in_ports().values() if not port.disconnected()]
+        num_inputs = len(connected_in_ports)
+        assert node.has_valid('equation'), "Einsum node {} must contain `equation` attribute".format(node_name)
+        equation = node.equation
+
+        # parse the equation and extract input and output subscripts
+        input_subscripts, output_subscript = Einsum.parse_equation(node_name, equation)
+
+        # check that each operand has the corresponding input subscript
+        assert len(input_subscripts) == num_inputs, "The number of input operands of Einsum node {} " \
+                                                    "must match the number of input subscripts " \
+                                                    "in `equation`".format(node_name)
+
+        # check compatibility of dimension sizes with the same label and generate a dictionary of shapes for labels
+        label_to_shape = {}
+        for input_ind in range(num_inputs):
+            input_shape = node.in_port(input_ind).data.get_shape()
+            input_subscript = input_subscripts[input_ind]
+            labels = Einsum.extract_subscript_labels(node_name, input_subscript)
+            num_dims = len(input_shape)
+            num_labels = len(labels)
+            dim_ind = 0
+            for label_ind in range(num_labels):
+                label = labels[label_ind]
+                num_broadcasted_dims = num_dims - num_labels + 1
```

Comment:
may be need to check dim_ind < len(input_shape)

Reply:
Done.